presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java

```diff
@@ -284,6 +277,31 @@ protected Page computeNext()
         return types;
     }
 
+    /**
+     * Update memory usage with extra memory needed.
+     * The returned value indicates whether the reservation is within the limit
+     */
+    private boolean updateMemoryWithReservation(long reservedMemoryBytes)
+    {
+        long memorySize = getSizeInMemory() + reservedMemoryBytes;
```

Comment:
I don't like how this code is reentrant. `GroupByHash` in the middle of mutating work calls some memory reservation callback. This can be fine, but the callback calls  `GroupByHash` again. It's not obvious that it sees the state it is supposed to see (before any changes?) or that it sees consistent state at all. We're mutating the state after all.

I see two possible ways out of this:
* make the callback accept full memory reservation rather than delta. This eliminates reentrancy 
* let the `GroupByHash` manage its memory directly (perhaps something like `AggregatedMemoryContext` but for non-system memory)

Reply:
Second option is closer to what we do in #8800 for many other parts of the code.

Reply:
I agree with @findepi about the reentrancy. I personally prefer the second alternative @findepi proposed. I'm ok with this as long as this is temporary. Add TODO?